Lib/fractions.py

```diff
@@ -305,6 +306,28 @@ def __new__(cls, numerator=0, denominator=None):
         self._denominator = denominator
         return self
 
+    @classmethod
+    def from_number(cls, number):
```

Comment:
Hmm, you can reuse this in the Rational constructor, isn't?  First three if/elif's share exactly same logic.

Reply:
Do you mean the main Fraction constructor?

There is a different exception if the argument type is not supported. There is also a matter of performance.

Reply:
I meant some helper:
```py
@classmethod
def _from_number(cls, number):
    if type(number) is int:
        return cls._from_coprime_ints(number, 1)
    elif isinstance(number, numbers.Rational):
        return cls._from_coprime_ints(number.numerator, number.denominator)
    elif (isinstance(number, float) or
          (not isinstance(number, type) and
           hasattr(number, 'as_integer_ratio'))):
        return cls._from_coprime_ints(*number.as_integer_ratio())
```

Then you can reuse one in the constructor and in the ``from_number()``, e.g.:
```
        if denominator is None:
            self = cls._from_number(numerator)
            if self is not None:
                return self

            elif isinstance(numerator, str):            
                ...
```

> There is also a matter of performance.

I don't expect too much from extra class method call.

Reply:
Extra function call and extra check.

This will not save much code. I'll leave this on future if we touch this code again.

Reply:
Well, it's up to you.  PR has Mark approval, so it's ready to go.